contracts/Pool.sol

```diff
@@ -137,36 +132,25 @@ contract Pool is Parameters, Initializable, IERC20Receiver, Ownable {
                 require(success);
             }
 
-        } else if (_tx_type()==3) { // Bridge deposit
+        } else if (_tx_type()==3) { // Permittable token deposit
             require(token_amount>=0 && energy_amount==0 && msg.value == 0, "incorrect deposit amounts");
-            require(bridge_deposits[_deposit_spender()] >= uint256(token_amount) * denominator, "not enough tokens for deposit");
-            bridge_deposits[_deposit_spender()] -= uint256(token_amount) * denominator;
+            (uint8 v, bytes32 r, bytes32 s) = _permittable_deposit_signature();
+            address holder = _memo_permit_holder();
+            token.permit(
+                holder,
+                address(this),
+                uint256(token_amount) * denominator,
+                _memo_permit_deadline(),
+                v,
+                r,
+                s
+            );
+            token.safeTransferFrom(holder, address(this), uint256(token_amount) * denominator);
         } else revert("Incorrect transaction type");
 
         if (fee>0) {
-            wxdai.safeTransfer(msg.sender, fee*denominator);
+            token.safeTransfer(msg.sender, fee*denominator);
```

Comment:
alternatively
this 
```
token.safeTransferFrom(holder, address(this), uint256(token_amount) * denominator);
```
should be changed to

```
token.safeTransferFrom(holder, address(this), uint256(_transfer_token_amount()) * denominator);
```
to exclude double fee transfer

Reply:
This is a fee that is sent to relayer from Pool contract. So, as a deposit Pool takes from the user `deposit_amount+fee`, then `fee` is used to pay the relayer

Reply:
@popov-vladimir here is the idea of the deposit logic in the contract:
1. get funds (deposit amount + fee) from the user account to the pool contracts (safeTransferFrom). 
2. transfer fees to the relayer account (safeTransfer)

Finally only the deposit amount will be locked on the pool contract.

This logic works for both deposits: the deposit with prior allowance (tx_type=0) and the deposit with the permit call (tx_type=3).

 Moreover, fees should be payed for other tx types: for transfers and for withdrawals. That's why the fees transfer is executed unconditionally.

Reply:
Ok, but than 

> This is a fee that is sent to relayer from Pool contract. So, as a deposit Pool takes from the user `deposit_amount+fee`, then `fee` is used to pay the relayer

Yeah, but it charges the client twice, which overflows the allowance.
Right now any positive fee will make the transaction revert

Reply:
> Yeah, but it charges the client twice, which overflows the allowance.
are you sure? As I mentioned above the first transfer (safeTransferFrom) is to get token from the account, the token will be on the pool account after that. So, the seconds transfer (safeTransfer) is to send tokens from the pool account to msg.sender. The second transfer does not deal with the allowance provided by the user for the first transfer at all.

Reply:
By the way, the transfer transaction (tx_type=1) with fee will definitely reverted due to [this require](https://github.com/zkBob/pool-evm-single-l1/blob/1e6d3d1f01bd0acd90a1da0e7983c810775d146f/contracts/Pool.sol#L117). Because `token_amount` will be greater than zero. 

I don't see any problems with deposit, withdrawal and permit-deposit transactions

Reply:
> By the way, the transfer transaction (tx_type=1) with fee will definitely reverted due to [this require](https://github.com/zkBob/pool-evm-single-l1/blob/1e6d3d1f01bd0acd90a1da0e7983c810775d146f/contracts/Pool.sol#L117). Because `token_amount` will be greater than zero.

When client builds a transfer tx with `fee>0,` one should specify `amount=-fee`, so that [this sum](https://github.com/zkBob/pool-evm-single-l1/blob/1e6d3d1f01bd0acd90a1da0e7983c810775d146f/contracts/Pool.sol#L110) is equal to 0. Basically, client pays the fee from assets that are already in the pool

Reply:
> By the way, the transfer transaction (tx_type=1) with fee will definitely reverted due to [this require](https://github.com/zkBob/pool-evm-single-l1/blob/1e6d3d1f01bd0acd90a1da0e7983c810775d146f/contracts/Pool.sol#L117). Because token_amount will be greater than zero.

Oh, sorry, it's my mistake. I have checked the libzeropool-rs code and in this case the `_transfer_token_amount()` will be equal `-fee`

Reply:
> When client builds a transfer tx with fee>0, one should specify amount=-fee, so that [this sum](https://github.com/zkBob/pool-evm-single-l1/blob/1e6d3d1f01bd0acd90a1da0e7983c810775d146f/contracts/Pool.sol#L110) is equal to 0. Basically, client pays the fee from assets that are already in the pool

The only thing I don't like is that we can guess about transferred amount based on fee amount. Especially if it is a default % in the UI.

Reply:
> The only thing I don't like is that we can guess about transferred amount based on fee amount. Especially if it is a default % in the UI.

Why fee amount should depends on transfer amount?

Reply:
> The only thing I don't like is that we can guess about transferred amount based on fee amount. Especially if it is a default % in the UI.

Yeah, if fee is deducted based on the amount transferred, then it is problem. It should be constant, just to cover relayer expenses

Reply:
Or it can be a % of the amount in case of deposit/withdrawal if we want so

Reply:
But why? Your transaction fee on mainnet never depends on the amount. Why it should calculate from the transfer amount in our solution? I think it's not so fair to users

Reply:
Yes, for me it looks OK to have it constant.

Reply:
I have just added fee in the current test script. It's seems works well https://github.com/zkBob/pool-evm-single-l1/pull/10/commits/2446cefa8d1e815878463a324cbfb07de74bfe11

Reply:
> > Yeah, but it charges the client twice, which overflows the allowance.
> 
> are you sure? As I mentioned above the first transfer (safeTransferFrom) is to get token from the account, the token will be on the pool account after that. So, the seconds transfer (safeTransfer) is to send tokens from the pool account to msg.sender. The second transfer does not deal with the allowance provided by the user for the first transfer at all.

@LyzeOfKiel @akolotov 
Actually, I was wrong, sorry for the mess.
The code is Ok,  and the test works fine if the deposit value is chosen accordingly with the fee amount